### PR TITLE
Do not emit duplicate ranges

### PR DESCRIPTION
**Pull request by efritz**
It turns out that just following `package.Syntax` will have you visit the same document multiple times.

After this PR, the order will be:

- emit all document elements
- emit all def elements
- emit all ref elements
- emit all contains and item edges

We do this so that we can have a set of "whitelisted" documents while emitting document elements. This ensures that the output does not have documents from generated code (go test will create a slimmed down version of the code and re-index it).

We then maintain the set of filenames that we've emitted defs and uses for so we only visit them the first time in the sequence.

Lastly, we maintain a map of (filename, offset) pairs so that we do not emit a duplicate range, which happens mostly when a def and use shares an offset in a file. We can then just use the same range here.

After this change the output passes our validation util. As we do emit many fewer elements than before, I am going to re-benchmark what's on our LSIF documentation to see if there's a significant benefit here.

### CODE REVIEW

File: index/indexer.go [hunk 1]
```diff
@@ -204,35 +235,71 @@ func (e *indexer) index(pkgs []*packages.Package) (*Stats, error) {
...
 
-func (e *indexer) indexPkg(p *packages.Package, proID string) (err error) {
-	log.Infoln("Package:", p.Name)
+func (e *indexer) indexPkgDocs(p *packages.Package, proID string) (err error) {
+	log.Infoln("Emitting documents for package", p.Name)
 	defer log.Infoln()
 
 	for _, f := range p.Syntax {
 		fpos := p.Fset.Position(f.Package)
+		if !strings.HasPrefix(fpos.Filename, e.projectRoot) {
```
Comment (unknwon) [inline index/indexer.go]: Not a blocker, but feels like a regression? https://github.com/sourcegraph/lsif-go/issues/27

Comment (unknwon) [reply index/indexer.go]: Should have commented on the next line:

>// Indexing test files means that we're also indexing the code generated by go test;

Comment (efritz) [reply index/indexer.go]: These files won't ever be served by the LSIF backend. I will verify by hand that test files still have support (that was originally enabled by setting `Test:true` when loading the package).

Comment (unknwon) [reply index/indexer.go]: Maybe just the comment itself confused me 😁

Comment (efritz) [reply index/indexer.go]: Still indexes test files.

**Pull request merged**